### PR TITLE
[OSJC-184] Add port forwarding capability

### PR DESCRIPTION
**Pull request by jcantrill**
@adietish @maxandersen This is for your initial review.  I'll push the JBT code that depends upon this.  I need to do some more testing for cases like when it cant find the binary and also on windows.

### CODE REVIEW

**Comment by adietish** [comment]
looks fine to me. Like the capability visitor.

File: src/main/java/com/openshift/internal/restclient/capability/resources/OpenShiftBinaryPortForwarding.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+	@Override
+	public synchronized void forwardPorts(PortPair... pairs) {
+		stop();
+		if(pairs.length == 0) return;
+		String location = System.getProperty(IPortForwardable.OPENSHIFT_BINARY_LOCATION);
+		if(StringUtils.isBlank(location)) {
+			throw new LocationNotFoundException(String.format("The OpenShift 'oc' binary location was not specified. Set the property %s", IPortForwardable.OPENSHIFT_BINARY_LOCATION));
+		}
+		StringBuilder args = new StringBuilder(location);
```
Comment (adietish) [inline src/main/java/com/openshift/internal/restclient/capability/resources/OpenShiftBinaryPortForwarding.java]: I would extract the path building to a separate method, that would keep this method focused and enhance readability

File: src/main/java/com/openshift/internal/restclient/capability/resources/OpenShiftBinaryPortForwarding.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+			.append("--server=").append(client.getBaseURL()).append(" ");
+			addToken(args)
+			.append("-n ").append(pod.getNamespace()).append(" ")
+			.append("-p ").append(pod.getName()).append(" ");
+		for (PortPair pair : pairs) {
+			args.append(pair.getLocalPort()).append(":").append(pair.getRemotePort()).append(" ");
+		}
+		ProcessBuilder builder = new ProcessBuilder(StringUtils.split(args.toString(), " "));
+		LOG.debug("PortForwarding args: {}", builder.command());
+		try {
```
Comment (adietish) [inline src/main/java/com/openshift/internal/restclient/capability/resources/OpenShiftBinaryPortForwarding.java]: maybe launching the process could make it to a separate method, keeping the calls in this method to something very readable like:
startProcess(createCommandLine(PortPair... pairs))

**Comment by adietish** [comment]
I tried to build my local origin so that I have the oc available, but it failed for me: http://fpaste.org/237582/55745821/
any idea what's wrong for me?

**Comment by jcantrill** [comment]
@adietish try 'make clean; make'

**Comment by jcantrill** [comment]
@adietish refactored and made more readable per your suggestions.  Still needs testing on windows and mac

**Comment by adietish** [comment]
@jcantrill  ok that helped, thx!

**Comment by adietish** [comment]
+1 good to go. can you merge and push?

**Comment by adietish** [comment]
I have a test failure when running the tests via maven:

```
Tests in error: 
   getContainerPorts(com.openshift.internal.restclient.model.v1beta3.PodTest): No path was found for property 'ports.name' in the property map.
```

**Comment by jcantrill** [comment]
Fixed test and repushed

**Comment by jcantrill** [comment]
@adietish I have tested on linux and windows and am comfortable with merging this.  Please review and comment accordingly.

**Comment by jcantrill** [comment]
merged into master

**Pull request closed without merging**